[qr]: Add support for `?qr`

erbesharat opened the pull request:
**What this PR does / why we need it**:
Adds support for `?qr`

**Which issue this PR fixes**:
fixes #258

stp-ip commented on setup_test.go:
```diff
@@ -382,6 +382,51 @@ func TestCaddyParse(t *testing.T) {
+		{
+			`
+			txtdirect {
+				enable host qr
```
In my view enable should only be used to enable routing/type features not general features. Aka we don't enable prometheus, but we do enable host and path.

So enable qr shouldn't be needed.

stp-ip commented on qr.go:
```diff
@@ -0,0 +1,87 @@
+	}
+	Qr.BackgroundColor, Qr.ForegroundColor = qr.backgroundColor, qr.foregroundColor
+
+	Qr.Write(qr.Size, w)
```
We should write a minimal html file instead. In Firefox for example a default image written back is transparent and therefore makes it hard to scan. Also helpful would be to add a clickable link to the actual webpage under the image so one can navigate there, if the qr code isn't necessary.

stp-ip replied:
I can see the reason to not add html. It would make for easier embedding.

Let's discuss this before implementing a change.

erbesharat replied:
Like the clickable link idea. But can't understand what you mean by easier embedding.

For the Firefox issue, we can add a custom white background color based on the request's `User Agent` header if writing HTML can't be considered.

stp-ip replied:
Any reason not to default to having a custom white background color in general?

stp-ip replied:
For context on the embedding.
Having the URL give back an image makes embedding into a different site etc. easier as it's just an image instead of playing with an iframe, when a full blown html would be given back. Does that make sense?

erbesharat replied:
I feel like dealing with iframe just makes the embedding harder. Think the image with a colored background should be enough and the user can add a link to the image while embedding it if they want it clickable.

stp-ip commented on qr.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (qr *Qr) ParseQr(c *caddy.Controller) error {
+	switch c.Val() {
```
If we go with being able to configure these things (a lot of configuration options in the long run) we should expose the error recovery.

stp-ip commented:
Also needs a rebase as I was too slow in reviewing :(

stp-ip commented:
So 2 things missing before merge:
* [ ] expose error recovery options for generated qr codes?
* [ ] custom white background?

Happy to discuss these.

erbesharat commented:
Seems like that we already have a default white background when the background color is empty in the config.
https://github.com/txtdirect/txtdirect/blob/8644cbd3b9abfc06604f299c34cb75d57be34ff2/qr.go#L75-L77

btw, I couldn't understand what you mean by exposing error recovery. can you please elaborate?

stp-ip commented:
The library has a specific value for error recovery. It makes the QR code
have more data, but be easier to recover.

On Sat, Aug 17, 2019, 1:11 PM Erfan Besharat <notifications@github.com>
wrote:

> Seems like that we already have a default white background when the
> background color is empty in the config.
>
> https://github.com/txtdirect/txtdirect/blob/8644cbd3b9abfc06604f299c34cb75d57be34ff2/qr.go#L75-L77
>
> btw, I couldn't understand what you mean by exposing error recovery. can
> you please elaborate?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/txtdirect/txtdirect/pull/269?email_source=notifications&email_token=AA4YAXEWM52FGWREC5UDNLLQE7MH3A5CNFSM4HR24T22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4QJC2A#issuecomment-522228072>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4YAXAQHFFVTSAF3ONGQMTQE7MH3ANCNFSM4HR24T2Q>
> .
>

stp-ip commented:
Yeah let's go with image. The problem with the white background is that
it's transparent I think. At least when I tested in Firefox it was
transparent.

On Sat, Aug 17, 2019, 2:43 PM Erfan Besharat <notifications@github.com>
wrote:

> *@erbesharat* commented on this pull request.
> ------------------------------
>
> In qr.go
> <https://github.com/txtdirect/txtdirect/pull/269#discussion_r314945968>:
>
> > +
> +	backgroundColor color.Color
> +	foregroundColor color.Color
> +}
> +
> +func (qr *Qr) Redirect(w http.ResponseWriter, r *http.Request) error {
> +	Qr, err := qrcode.New(r.Host+r.URL.String(), qrcode.High)
> +	if err != nil {
> +		return fmt.Errorf("Couldn't generate the Qr instance: %s", err.Error())
> +	}
> +	if err = qr.ParseColors(); err != nil {
> +		return fmt.Errorf("Coudln't parse colors: %s", err.Error())
> +	}
> +	Qr.BackgroundColor, Qr.ForegroundColor = qr.backgroundColor, qr.foregroundColor
> +
> +	Qr.Write(qr.Size, w)
>
> I feel like dealing with iframe just makes the embedding harder. Think the
> image with a colored background should be enough and the user can add a
> link to the image while embedding it if they want it clickable.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/txtdirect/txtdirect/pull/269?email_source=notifications&email_token=AA4YAXD7UTW63SR63LETZGDQE7W5PA5CNFSM4HR24T22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCB3QNGI#discussion_r314945968>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4YAXBDWWBPDCKPX5Z333DQE7W5PANCNFSM4HR24T2Q>
> .
>

erbesharat commented:
The foreground was transparent, maybe that was the problem in Firefox. Changed the default foreground color and added recovery_level option to config.

stp-ip commented:
/lgtm

On Sun, Aug 18, 2019, 2:06 PM Erfan Besharat <notifications@github.com>
wrote:

> The foreground was transparent, maybe that was the problem in Firefox.
> Changed the default foreground color and added recovery_level option to
> config.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/txtdirect/txtdirect/pull/269?email_source=notifications&email_token=AA4YAXHRYADVAYZU5ZAFQBDQFE3MNA5CNFSM4HR24T22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4Q6RLQ#issuecomment-522315950>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA4YAXDCW4YXFY5MCIPJPQDQFE3MNANCNFSM4HR24T2Q>
> .
>
